certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -616,26 +596,48 @@ def _enable_redirect(self, domain, unused_options):
         vhost = None
         # If there are blocks listening plaintextishly on self.DEFAULT_LISTEN_PORT,
         # choose the most name-matching one.
+
+        # TODO: should we be adding redirects to all matching blocks?
+
         vhost = self.choose_redirect_vhost(domain, port)
 
         if vhost is None:
             logger.info("No matching insecure server blocks listening on port %s found.",
                 self.DEFAULT_LISTEN_PORT)
+            return
+
+        if vhost.ssl:
+            new_vhost = self.parser.duplicate_vhost(vhost,
+                only_directives=['listen', 'server_name'])
```

Comment:
As an aside, if we do this, we may want to do something special with our old `if` style redirect. Otherwise, for existing Certbot users who ask again for a redirect, we'll somewhat awkwardly say there's an existing redirect and insert a comment.

We could get around this either by continuing to check for the old style redirect or looking for `# managed by Certbot` comments.

Reply:
The reason I did this is if there's an existing rewrite in the original block. If there is, this way will have http redirect to https, which then gets whatever rewrite they originally wanted. If we leave it, then we won't stick the redirect in there at all.

Reply:
The log thing is a good point though.

Reply:
And also, as you noticed, it lets us not have to deal with the awkward old-redirect thing going on, and just completely remove that from the codebase as irrelevant.

Reply:
I'd almost prefer to explicitly copy over things like logs, and say that if it's anything more complicated than that, should they really be letting us put in redirects for them anyway?

Reply:
> The reason I did this is if there's an existing rewrite in the original block. If there is, this way will have http redirect to https, which then gets whatever rewrite they originally wanted. If we leave it, then we won't stick the redirect in there at all.

I see what you're saying, but I think this is actually a little tricky (actually more so than when I initially reviewed this PR). For example, imagine the user has the following in their server block:
```
if ($host = "example.org") {
    return 301 http://www.example.org$request_uri;
}
```
When we go to add the redirect, we remove these lines and add a redirect to the HTTPS version of the site. Now there is a redirect loop.

I think that while it'd be nice to remove existing redirect directives to automatically handle cases like the one you described, I think doing so is very difficult. Additionally, I think that in the case we're creating a HTTP server block from an HTTPS one, we should refuse to add a redirect if one exists in the HTTPS block to avoid problems like this.

> I'd almost prefer to explicitly copy over things like logs, and say that if it's anything more complicated than that, should they really be letting us put in redirects for them anyway?

As for whether or not to copy directives in general, while I don't have a very strong opinion, I think I'd rather err the other way. In each case, what's the worst that can happen?

If we don't copy directives, we'll remove configuration for things like logs that were already set up. I feel like if we make a list of directives to copy over, we'll constantly be adding to it as new bug reports come in as we can't possibly know about all crazy directives out there.

If we do copy everything over and refuse to add a redirect when one exists in either the HTTP or HTTPS server block, while we leave a lot of unnecessary stuff in their server block, I'm unaware of any possibilities of incorrect behavior from nginx here. It seems like even if we leave things like SSL directives in the block, Nginx just ignores them.

Reply:
I suppose the example I gave isn't exactly a redirect loop. Worst case is:

http://example.org -> https://example.org -> http://www.example.org -> https://www.example.org

This is obviously less than ideal though and I'm not convinced it's impossible for us to create redirect loops with this approach.

Reply:
1.
> If we do copy everything over and refuse to add a redirect when one exists in either the HTTP or HTTPS server block, while we leave a lot of unnecessary stuff in their server block, I'm unaware of any possibilities of incorrect behavior from nginx here. It seems like even if we leave things like SSL directives in the block, Nginx just ignores them.

Yes, that's what happens.

2. Your example wouldn't actually cause a redirect loop.

Let's say they had:
```
server {
  server_name example.org www.example.org;
  if ($host = "example.org") {
      return 301 http://www.example.org$request_uri;
  }
}
```
Then, we deploy:
```
server {
  server_name example.org www.example.org;
  if ($host = "example.org") {
      return 301 http://www.example.org$request_uri;
  }
  listen 80;
  listen 443 ssl; etc.
}
```
Now we add a redirect for example.org:
```
server {
  server_name example.org www.example.org;
  listen 80;
  return 301 https://$host$request_uri;
}
server {
  server_name example.org www.example.org;
  if ($host = "example.org") {
      return 301 http://www.example.org$request_uri;
  }
  listen 443 ssl; etc.
}
```
Now we add a redirect for www.example.org:
```
server {
  server_name example.org www.example.org;
  listen 80;
  return 301 https://$host$request_uri;
}
server {
  server_name example.org www.example.org;
  listen 80;
  return 301 https://$host$request_uri;
}
server {
  server_name example.org www.example.org;
  if ($host = "example.org") {
      return 301 http://www.example.org$request_uri;
  }
  listen 443 ssl; etc.
}
```
(I admit that this is superfluous, and we should in fact be just putting the single requested server_name per newly created server block.)
This isn't a redirect loop. If you request example.org, it tells you to go to http://www.example.org, so you do, and that tells you to go to https://www.example.org. No problem here.

The only way it would cause a loop is if it there's a redirect from https to http in there with nothing else changed, which is not a thing that happens. That's why I decided to only keep those directives in the first place.

I don't think we should refuse to do this if there's a redirect in the existing server block. The nice thing about doing it the way I have implemented is it allows us to put in the https redirect for the common case illustrated above.

Even for the logs, they'll still get all the previous functionality when the request arrives on 443 one redirect later. If they want more reporting or fancier things to happen when 80 comes in, they can add those manually. This is the safest way that allows us to handle most of the cases.

Reply:
OK. I'm on board with adding redirects even if they exist in the HTTPS server block.

I still think we should keep other directives though. Why not just remove `redirect`, `return`, SSL `listen` statements, and ideally all code that is `# managed by Certbot`? What are you worried about from other directives?

Reply:
This weirdly went as a general reply rather than here, so here it is in its proper place:

Mainly if and include -- if we parse recursively and find something farther down, do we remove the entire highest level directive? Just the offending one? That gets messy. Plus, isn't the reason we're splitting it later to not have two blocks with duplicated directives they have to manage? And from what I've seen, the usual setup is just a redirect in one block and all the actual management code in the other. Which is what this would currently create.

Reply:
Also, opinion on doing the thing I said about just putting the single requested `server_name` per newly created server block?

Reply:
Note: this actually only creates a single new block.